### PR TITLE
fix(updater): run cleanup before exit on Windows

### LINKED ISSUE #1014: [updater] doesn't shutdown app gracefully on Windows
This skips app clean up code, for example kill sidecar / save windows states / save store

https://github.com/tauri-apps/plugins-workspace/blob/99bea2559c2c0648c2519c50a18cd124dacef57b/plugins/updater/src/updater.rs#L589

### PR DESCRIPTION
**Pull request by Legend-Master**
Fix #1014

### CODE REVIEW

**Comment by Legend-Master** [comment]
The downside about this is that we will have to assume `ShellExecute` will run successfully, since we need to clean up before the installer gets executed (installer will kill all running instances)

**Comment by Legend-Master** [comment]
And, I'm not familiar with how updaters shutdown the app on other platforms, is this just a Windows problem or we need to do something for other platforms as well?

**Comment by amrbashir** [comment]
> And, I'm not familiar with how updaters shutdown the app on other platforms, is this just a Windows problem or we need to do something for other platforms as well?

correct, other platforms update without restarting and wait for users to relaunch.

**Review by amrbashir — CHANGES_REQUESTED** [review]
I don't quite like the additional argument, I think a better way to handle this would be to take a generic callback to be executed on before restart, here is some psuedo-code of what I mean:
```rs
let builder = UpdaterBuilder::new()
  .with_before_exit(|| {
    app.cleanup_before_exist();
  })
  .build();
```

and we can provide this callback automatically in the `UpdaterExt` trait implementation.

**Comment by Legend-Master** [comment]
A bit off topic for this PR, do we want to expose `download` and `install` directly, so that the app can download the update in background and prompt the user once it's done (like the way vscode did it)?

**Comment by amrbashir** [comment]
they are exposed https://docs.rs/tauri-plugin-updater/latest/tauri_plugin_updater/struct.Update.html#method.download

**Comment by Legend-Master** [comment]
> I think a better way to handle this would be to take a generic callback to be executed on before restart

I'll take a look

> they are exposed https://docs.rs/tauri-plugin-updater/latest/tauri_plugin_updater/struct.Update.html#method.download

I mean the JavaScript side

**Comment by amrbashir** [comment]
> I mean the JavaScript side

yeah sure, was hoping to do it at some point

File: .changes/fix-updater-cleanup-md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"updater": patch
+---
+
+Add a `on_before_exit` hook for cleanup before spawning the updater on Windows, defaults to `app.cleanup_before_exit`
```
Comment (amrbashir) [inline .changes/fix-updater-cleanup-md]:
```suggestion
Add a `on_before_exit` hook for cleanup before spawning the updater on Windows, defaults to `app.cleanup_before_exit` when used through `UpdaterExt`
```

**Review by amrbashir — DISMISSED** [review]
Thank you

**Review by amrbashir — DISMISSED** [review]

File: plugins/updater/src/updater.rs [hunk 2]
```diff
@@ -375,9 +388,10 @@ impl Updater {
     }
 }
 
-#[derive(Debug, Clone)]
+#[derive(Clone)]
 pub struct Update {
     config: Config,
+    on_before_exit: Option<OnBeforeExit>,
```
Comment (amrbashir) [inline plugins/updater/src/updater.rs:395]:
```suggestion
    #[allow(unused)]
    on_before_exit: Option<OnBeforeExit>,
```

**Review by amrbashir — DISMISSED** [review]

**Review by amrbashir — APPROVED** [review]

**Pull request merged**